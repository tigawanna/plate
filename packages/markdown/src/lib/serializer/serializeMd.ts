import type { Descendant, SlateEditor } from 'platejs';

import remarkStringify from 'remark-stringify';
import { type Plugin, unified } from 'unified';

import type { AllowNodeConfig } from '../MarkdownPlugin';
import type { MdRoot } from '../mdast';
import type { MdRules, PlateType } from '../types';

import { convertNodesSerialize } from './convertNodesSerialize';
import { getMergedOptionsSerialize } from './utils/getMergedOptionsSerialize';
export type SerializeMdOptions = {
  allowedNodes?: PlateType[] | null;
  allowNode?: AllowNodeConfig;
  disallowedNodes?: PlateType[] | null;
  editor?: SlateEditor;
  preserveEmptyParagraphs?: boolean;
  remarkPlugins?: Plugin[];
  rules?: MdRules;
  spread?: boolean;
  value?: Descendant[];
  withBlockId?: boolean;
};

/** Serialize the editor value to Markdown. */
export const serializeMd = (
  editor: SlateEditor,
  options?: Omit<SerializeMdOptions, 'editor'>
) => {
  const mergedOptions = getMergedOptionsSerialize(editor, options);

  const { remarkPlugins, value } = mergedOptions;

  const toRemarkProcessor = unified()
    .use(remarkPlugins ?? [])
    .use(remarkStringify, {
      emphasis: '_',
    });

  const mdast = slateToMdast({
    children: value!,
    options: mergedOptions,
  });

  return toRemarkProcessor.stringify(mdast);
};

const slateToMdast = ({
  children,
  options,
}: {
  children: Descendant[];
  options: SerializeMdOptions;
}): MdRoot => {
  const processedChildren = convertNodesSerialize(
    children,
    options,
    true // isBlock = true for top-level elements
  ) as MdRoot['children'];

  const ast = {
    children: processedChildren,
    type: 'root',
  } as MdRoot;
  return ast;
};
