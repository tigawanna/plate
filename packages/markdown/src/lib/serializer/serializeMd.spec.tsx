/** @jsx jsxt */

import { jsxt } from '@platejs/test-utils';

import { createTestEditor } from '../__tests__/createTestEditor';
import { testValue } from '../__tests__/testValue';
import { serializeMd } from './serializeMd';
const editor = createTestEditor();

jsxt;

describe('serializeMd', () => {
  it('should serialize a simple paragraph', () => {
    const result = serializeMd(editor as any, {
      spread: true,
      value: testValue,
    });
    expect(result).toMatchSnapshot();
  });

  describe('fixures', () => {
    // https://github.com/inokawa/remark-slate-transformer/issues/44
    it('should serialize marks with trailing whitespace', () => {
      const slateNodes = [
        {
          children: [
            {
              text: 'Make text ',
            },
            {
              bold: true,
              text: 'bold',
            },
            {
              text: ', ',
            },
            {
              italic: true,
              text: 'italic',
            },
            {
              text: ', ',
            },
            {
              text: 'underlined',
              underline: true,
            },
            {
              text: ', or apply a ',
            },
            {
              bold: true,
              italic: true,
              text: 'combination',
              underline: true,
            },
            {
              text: ' of these styles for a visually striking effect.',
            },
          ],
          type: 'p',
        },
      ];
      const result = serializeMd(editor as any, { value: slateNodes });
      expect(result).toMatchSnapshot();
    });

    // https://github.com/inokawa/remark-slate-transformer/issues/90
    it('Combination of bold and italic at the start of a block losing formatting', () => {
      const slateNodes = [
        {
          children: [
            {
              bold: true,
              italic: true,
              text: 'Italic',
            },
            {
              bold: true,
              text: ' in a bold paragraph',
            },
          ],
          type: 'p',
        },
        {
          children: [
            {
              bold: true,
              text: 'This is an ',
            },
            {
              bold: true,
              italic: true,
              text: 'Italic',
            },
            {
              bold: true,
              text: ' in a bold paragraph',
            },
          ],
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, { value: slateNodes });
      expect(result).toMatchSnapshot();
    });

    // https://github.com/inokawa/remark-slate-transformer/issues/145
    it('should serialize a value with inline code', () => {
      const slateNodes = [
        {
          children: [
            { bold: true, code: true, italic: true, text: 'Inline code' },
          ],
          type: 'p',
        },
        {
          children: [{ code: true, italic: true, text: 'Inline code' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it('should serialize a value within a block qoute as a single line', () => {
      const slateNodes = [
        {
          children: [
            { text: 'Block quote' },
            { text: ' with a new line ' },
            { bold: true, code: true, italic: true, text: ' Inline code' },
          ],
          type: 'blockquote',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it('should serialize a codeblock', () => {
      const slateNodes = [
        {
          children: [
            { text: 'Code block 1 line 1' },
            { text: 'Code block 1 line 2' },
            { children: [{ text: 'Code block 1 line 3' }], type: 'code_line' },
            { text: 'Code block 1 line 4', type: 'code_line' },
          ],
          type: 'code_block',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it(
      String.raw`should serialize a \n within a block qoute as new line`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Block quote' },
              { text: '\n' },
              { text: 'with a new line' },
            ],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );

    it(
      String.raw`should serialize two \n within a block qoute as two new lines`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Block quote' },
              { text: '\n' },
              { text: '\n' },
              { text: 'with a new line' },
            ],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );

    it(
      String.raw`should serialize two leading \n at the end of a block qoute as a new line`,
      () => {
        const slateNodes = [
          {
            children: [{ text: 'Block quote' }, { text: '\n' }, { text: '\n' }],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );

    it(
      String.raw`should serialize three leading \n at the end of a paragraph qoute as a new line`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Paragaph with two new Lines' },
              { text: '\n' },
              { text: '\n' },
              { text: '\n' },
            ],
            type: 'p',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );
  });

  it(
    String.raw`should serialize the leading break at the end of a block qoute as a <br />`,
    () => {
      const slateNodes = [
        {
          children: [{ text: 'Paragaph with a new Line' }, { text: '\n' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    }
  );

  it(
    String.raw`should serialize paragraphs with only a new line to a <br />`,
    () => {
      const slateNodes = [
        {
          children: [{ text: '\n' }],
          type: 'p',
        },
        {
          children: [{ text: '\n' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    }
  );

  it('should serialize lists with spread option correctly', () => {
    const listFragment = [
      {
        children: [{ text: '1' }],
        indent: 1,
        listStyleType: 'decimal',
        type: 'p',
      },
      {
        children: [{ text: '2' }],
        indent: 1,
        listStart: 2,
        listStyleType: 'decimal',
        type: 'p',
      },
    ];

    // Test with default (spread: false)
    const resultDefault = serializeMd(editor as any, { value: listFragment });
    expect(resultDefault).toBe('1. 1\n2. 2\n');

    // Test with spread: true
    const resultWithSpread = serializeMd(editor as any, {
      spread: true,
      value: listFragment,
    });
    expect(resultWithSpread).toBe('1. 1\n\n2. 2\n');
  });
});
