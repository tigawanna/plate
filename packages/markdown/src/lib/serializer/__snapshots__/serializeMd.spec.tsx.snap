// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeMd fixures Combination of bold and italic at the start of a block losing formatting 1`] = `
"_**Italic**_ **in a bold paragraph**

**This is an _Italic_** **in a bold paragraph**
"
`;

exports[`serializeMd fixures should serialize a \\n within a block qoute as new line 1`] = `
"> Block quote\\
> with a new line
"
`;

exports[`serializeMd fixures should serialize a codeblock 1`] = `
"\`\`\`
Code block 1 line 1
Code block 1 line 2
Code block 1 line 3
Code block 1 line 4
\`\`\`
"
`;

exports[`serializeMd fixures should serialize a value with inline code 1`] = `
"_**\`Inline code\`**_

_\`Inline code\`_
"
`;

exports[`serializeMd fixures should serialize a value within a block qoute as a single line 1`] = `
"> Block quote with a new line _**\` Inline code\`**_
"
`;

exports[`serializeMd fixures should serialize marks with trailing whitespace 1`] = `
"Make text **bold**, _italic_, <u>underlined</u>, or apply a _**<u>combination</u>**_ of these styles for a visually striking effect.
"
`;

exports[`serializeMd fixures should serialize three leading \\n at the end of a paragraph qoute as a new line 1`] = `
"Paragaph with two new Lines\\
\\ 
<br />
"
`;

exports[`serializeMd fixures should serialize two \\n within a block qoute as two new lines 1`] = `
"> Block quote\\
> \\
> with a new line
"
`;

exports[`serializeMd fixures should serialize two leading \\n at the end of a block qoute as a new line 1`] = `
"> Block quote\\ 
> <br />
"
`;

exports[`serializeMd should serialize a simple paragraph 1`] = `
"# Playground

## Comments and suggestions

Add <comment>comments to your content</comment> to provide additional conte1xt, <suggestion>insightssuggestions</suggestion>, or <comment>collaborate</comment> with others

### Heading

* 1

  * 2

    * 3

      * 4

| 1-1 | 1-2 |
| --- | --- |
| 2-1 | 2-2 |

> blockQuote

***

normal **bold**_italic_~~strikethrough~~\`code\`_**combine**_[link](https://example.com)$E=mc^2$

![test](https://images.unsplash.com/photo-1712688930249-98e1963af7bd?q=80\\&w=2070\\&auto=format\\&fit=crop\\&ixlib=rb-4.0.3\\&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D "test")

\`\`\`javascript
// Use code blocks to showcase code snippets
function greet() {
  console.info('Hello World!');
}
\`\`\`

$$
f(x)
$$

<toc />

* [x] Todo item

* Numbered list item

Centered text

Text with custom line height

<file name="sample.pdf" src="https://example.com/sample.pdf" />

<audio src="https://example.com/audio.mp3" />

<video src="https://example.com/video.mp4" />

<column_group layout="[50,50]">
  <column width="50%">
    Left column
  </column>

  <column width="50%">
    Right column
  </column>
</column_group>

Mention: [User](mention:User)

Date: <date>2024-01-01</date>

Text with <mark>highlight</mark> and <u>underline</u> and <kbd>keyboard</kbd> formatting

Text with <span style="color: rgb(252, 109, 38);">color</span> and <span style="background-color: rgb(252, 109, 38);">background</span> color

Superscript: x<sup>2</sup>Subscript: x<sub>2</sub>

<callout>
  callout
</callout>
"
`;

exports[`serializeMd should serialize paragraphs with only a new line to a <br /> 1`] = `
"
<br />


<br />
"
`;

exports[`serializeMd should serialize the leading break at the end of a block qoute as a <br /> 1`] = `
"Paragaph with a new Line
<br />
"
`;
